Initial Application API Support

rrao24 opened the pull request:
Dan,

This branch is likely not ready to be merged in to the main 2.0 preview branch yet. Could you please take a look at the code I've written thus far and let me know if I am on the right track?

Thank you very much.

rrao24 committed:
Added get method for application.js

rrao24 committed:
Add application.js

rrao24 committed:
Fix some linting errors

rrao24 committed:
Add semicolon and docs

dtolb commented on docs/api.md:
```diff
@@ -13,6 +15,40 @@
 
+<a name="Application"></a>
+
+## Application
```
Did you type this out manually? I think that this has been auto generated somehow. @nguyer ?

dtolb commented on lib/application.js:
```diff
@@ -0,0 +1,31 @@
+
+var Application = function (client) {
+	/**
+	 * Get information about the specified applications
```
nitpick comment... check alignment :(

rrao24 committed:
Change body to qs in params

rrao24 committed:
Update Application.js

Changed get to list, added a pagination test, included info about
properties at the bottom of the file

fuchsnj commented on lib/application.js:
```diff
@@ -0,0 +1,56 @@
+	/**
+	 * List the user's applications
+	 * @param {Object} params Parameters for filtering applications
+	 * @param {Number} [params.page] The specified page requested when querying
```
We do not want to allow people to use the page param. If someone needs the 2nd page, they should follow the 'next' link.
Using the page parameters has a significant performance impact on Catapult and we are starting to migrate to something else that is faster. If you use the 'next' link we can switch to a different paging method without you knowing.

rrao24 replied:
@dtolb Could I get some clarification for this? On http://ap.bandwidth.com/docs/rest-api/applications/ it says page is a supported parameter

(@fuchsnj asking for clarification because I was told that we wanted the sdk to have complete featured parody with the api)

dtolb replied:
@fuchsnj I did not know that. We will discuss and update here. Are we going to update the documentation across the site?

So now people will search by `size` and then follow the `next` link in the header?

dtolb replied:
@rrao24 here is an example on how we do this in [messages](https://github.com/bandwidthcom/node-bandwidth/blob/2.0-preview/lib/message.js#L80)

fuchsnj commented on lib/application.js:
```diff
@@ -0,0 +1,56 @@
+			qs     : params
+		})
+		.then(function (response) {
+			return response.body;
```
It would be great if this gave you a way to retrieve the next page of results. Possibly by returning some kind of "handle" object that contains data and a method to retrieve the next page

dtolb replied:
@fuchsnj how does [this look](https://github.com/bandwidthcom/node-bandwidth/blob/2.0-preview/lib/message.js#L80).

I almost think we should abstract that functionality out.. maybe to part of the client?

fuchsnj commented on lib/application.js:
```diff
@@ -0,0 +1,56 @@
+	 * @param {Object} params Parameters for filtering applications
+	 * @param {Number} [params.page] The specified page requested when querying
+	 * a list of applications
+	 * @param {Number} [params.size] The size of each page requested
```
We should probably document the max page size (I believe it is 1000)

fuchsnj commented on lib/application.js:
```diff
@@ -0,0 +1,56 @@
+		.then(function (response) {
+			return response.body;
+		})
+		.asCallback(callback);
```
What happens if you provide a callback and the functions fails?

rrao24 replied:
I believe that this is automatically handled for us using the .asCallback method.

rrao24 committed:
More functionality added to applications library

Tests not written yet

rrao24 committed:
Add Functionality to Application Class

Rewrote Application list() function to use links, added all methods from
API to applications class, wrote tests for each method

rrao24 committed:
Fix linting errors

rrao24 commented:
Could you guys take another look at this PR? I have added all the functionality to the Application class as specified in the API, accounted for pagination, and have written appropriate tests.

Thank you very much for your feedback.

rrao24 committed:
Remove dependency

rrao24 committed:
Remove unnecessary line

dtolb commented on lib/application.js:
```diff
@@ -0,0 +1,186 @@
+			var applicationListResponse = {
+				applications : response.body
+			};
+			var links = response.headers.link.split(",");
```
What happens if `response.headers.link` doesn't exist or is undefined

rrao24 replied:
For applications, two links are returned, a link to the first page of applications and one to the next page of applications (if it exists). It is guaranteed to at least have that one link which references the first page of applications.

dtolb commented on lib/application.js:
```diff
@@ -0,0 +1,186 @@
+				applications : response.body
+			};
+			var links = response.headers.link.split(",");
+			var nextLink = links[1];
```
What if sizeof links is only 1?

rrao24 replied:
If sizeof links is only 1, then `nextLink = links[1]` will be undefined and be caught by the following line.

fuchsnj replied:
It is not guaranteed that the next link will always be the 2nd link, also we will need this code in several places so we should probably make a function that grabs the next link. Some resources, such as message don't contain the 'first' link so you will actually need to check if the link matches.

I have some example code that does this in the Rust SDK if that helps: https://github.com/bandwidthcom/rust-bandwidth/blob/master/src/util.rs#L15

rrao24 replied:
Created a utility library to handle this.

rrao24 committed:
Fix more linting issues

rrao24 committed:
Documentation Fix

rrao24 committed:
Created a utility library for getting the next link and added appropriate tests

rrao24 committed:
Merge remote-tracking branch 'upstream/2.0-preview' into 2.0-preview

rrao24 committed:
Switch string.incluldes() to string.indexOf() for compatibility reasons

rrao24 committed:
Docs update

rrao24 committed:
More docs updating

illiriks commented on docs/api.md:
```diff
@@ -1,6 +1,10 @@
+<dt><a href="#Application">Application</a></dt>
+<dd></dd>
+<dt><a href="#Application">Application</a> : <code>Object</code></dt>
+<dd></dd>
```
This seems to be the source of the duplication. The thing that Application methods return is also an Application for jsdoc... Perhaps rename it ApplicationResponse?

rrao24 replied:
Fixed, thanks

rrao24 committed:
Should fix duplicate docs problem

rrao24 committed:
Fix Illegal trailing whitespace

fuchsnj commented on lib/application.js:
```diff
@@ -0,0 +1,180 @@
+			method : "POST",
+			body   : params
+		})
+		.then(function () {
```
an empty .then() does the same as not calling .then(). You can remove these 3 lines

rrao24 committed:
Change a comment which was causing tests to fail

JSDoc does not support format of @param {Array[...]}

dtolb commented:
I think something isn't set right for the documentation or I'm using the SDK wrong.

The readme has

``` javascript
client.Message.send({
    from : "+12345678901", // This must be a Catapult number on your account
    to   : "+12345678902",
    text : "Hello world."
})
.then(function(message) {
    console.log("Message sent with ID " + message.id);
})
.catch(function(err) {
    console.log(err.message);
});
```

As the correct way. BUT when I view the full API docs it looks something like this
![screen shot 2016-06-29 at 16 45 02](https://cloud.githubusercontent.com/assets/6805534/16468234/e178e856-3e18-11e6-9867-687fcda0d6be.png)

It's a bit confusing. any suggestions? @nguyer @fuchsnj @a-type

rrao24 commented:
@dtolb: Perhaps the client isn't being configured correctly? The following code works for me:

```
var CatapultClient = require("./node-bandwidth/index");
var clientObj = new CatapultClient({
    userId : 'useridhere',
    apiToken : 'apitokenhere',
    apiSecret : 'apisecrethere'
});

clientObj.Message.send({
    from : "+19104270488",
    to   : "somenumber",
    text : "Hello world."
})
.then(function(data) {
    console.log(data);
});
```

EDIT: I see what you're saying. I believe that the three required fields are from, to, and text, and the rest of the parameters are optional.

dtolb commented:
I'll look into this a bit more... nothing functionally wrong here.

Would like to see a new PR updating the [`messages.list` pagination](https://github.com/bandwidthcom/node-bandwidth/blob/2.0-preview/lib/message.js#L84) to this new method

@fuchsnj you good?

rrao24 committed:
Rewrite message.js to consume util.js

rrao24 commented:
@dtolb: I went ahead and updated the messages.list method in this PR as it was only roughly a 15 line change.

dtolb commented:
@rrao24 alright sweet. Once tests pass and one other person confirms, we're good.

@nguyer @fuchsnj @a-type @nbragdon

rrao24 committed:
Revert "Rewrite message.js to consume util.js"

This reverts commit caee4a80a92978a2f0793a6d2754f4c45d4db83a.

rrao24 commented:
Update: @dtolb I'm reverting the previous commit I made which rewrites message.js to use util.js. It appears that the list command will return a 'next' link even if the following page is empty. Will investigate this further and hopefully create a separate PR, but for now, application.js should be fully functional and ready to be merged.

EDIT: It appears that the previously written code did not account for this either

nbragdon commented on lib/application.js:
```diff
@@ -0,0 +1,180 @@
+			};
+			var nextLink = getNextLink(response);
+			if (nextLink) {
+				applicationListResponse.hasMore = true;
```
I would probably call this `hasNextPage`, that is more explicit than hasMore.

nbragdon commented on lib/application.js:
```diff
@@ -0,0 +1,180 @@
+			qs     : params
+		})
+		.then(function (response) {
+			var applicationListResponse = {
```
I know it's not required but I'd say defaulting all the variables you plan to add to this object is a good practice, so add `hasNextPage` and `getNextPage` and what those should be defaulted to should they not exist.

nbragdon commented on lib/util.js:
```diff
@@ -0,0 +1,25 @@
+ * @param {Object} response A response object returned from calling 'client.makeRequest'
+ * @returns A parsed version of the link to the subsequent page, or null if no such page exists.
+ */
+var getNextLink = function (response) {
```
Just curious why we wrote this ourselves instead of using a library? https://github.com/thlorenz/parse-link-header . It could possibly be because we don't want our sdk to have a lot of dependencies which is fair, but then we probably shouldn't use qs either and just stick to core node libraries.

nbragdon replied:
Since this function really only cares about the headers we should probably just pass that in instead of the response. Makes it slightly more agnostic.

dtolb replied:
parse-link-header is extremely lightweight. Only requires xtend which is like 10 lines of code.

rrao24 replied:
I've re-implemented this functionality using querystring instead of qs, which is a core node library: https://nodejs.org/api/querystring.html

nbragdon commented:
Can we change util to something more specific? Util is a very generic, kitchen sink kind of name that entices people to just throw stuff into it. Currently there is only 1 function in it called getNextLink. So maybe something to do with response parsing, or even more specific like header parsing.

rrao24 committed:
Added some default behavior to applicationResponse

Also renamed util.js to headerParsingLib.js

rrao24 committed:
Remove debugging statement

rrao24 committed:
Switch from qs to querystring

querystring is a core node module

rrao24 committed:
headerparsinglib now expects a header object instead of a response

dtolb commented:
@nbragdon how you feel about changes now?

nbragdon commented:
LGTM but did we consider using https://github.com/thlorenz/parse-link-header and just decide no? Just want to make sure it was at least considered.

dtolb commented:
@nbragdon @rrao24 @illiriks could we handle the `next-page-stuff` at the [client.js](https://github.com/bandwidthcom/node-bandwidth/blob/2.0-preview/lib/client.js) level? That way anything that is paginated just returns the correct header vs adding that for each endpoint.

Is that even possible? 

I have no issue with moving to [Parse link header](https://github.com/thlorenz/parse-link-header). Would like to get one more opinion.  @a-type ??

rrao24 commented:
@dtolb: In terms of the next-page-stuff, things could be handled in client like so:

```
function handleResponse(response) {
        if (response.statusCode !== 200 && response.statusCode !== 201) {
            var message = "";
            if (response.body) {
                message = response.body.message || "";
            }
            throw new UnexpectedResponseError(message, response.statusCode);
        }
        response.hasNextPage = false;
        response.getNextPage = function (nextCallback) {
            return Promise.reject("Next page does not exist.")
            .asCallback(nextCallback);
        };
        var nextLink = getNextLink(response.headers);
        if (nextLink) {
            response.hasNextPage = true;
            response.nextLink = nextLink;
        }
        return response;
    }
```

And in Application:

```
    this.list = function (params, callback) {
        var self = this;
        return client.makeRequest({
            path   : "applications",
            method : "GET",
            qs     : params
        })
        .then(function (response) {
            var applicationListResponse = {
                applications : response.body,
                hasNextPage  : response.hasNextPage,
                getNextPage  : response.getNextPage
            };
            if (response.hasNextPage) {
                applicationListResponse.getNextPage = function (nextCallback) {
                    return self.list(response.nextLink, nextCallback);
                };
            }
            return applicationListResponse;
        })
        .asCallback(callback);
    };
```

Thoughts on this? It seems a bit clunkier to me, and we still have to do some handling in application.js, so it somewhat defeats the purpose, but maybe there are some improvements you may suggest?

Will look into using Parse link header if we feel that is the path we want to go down.

rrao24 committed:
Rewrite the headerparsinglib using parse-link-header

rrao24 commented:
The latest commit rewrites headerparsinglib.js using the parse-link-header library.
